### PR TITLE
[ci] Add code coverage github actions workflow

### PR DESCRIPTION
**Pull request by notmandatory**
The new code coverage workflow is run when a PR is merged to master. I'm using tarpaulin for code coverage data collection, and codecov.io for storing and displaying code coverage data.

This PR also adds a corresponding code coverage badge to README.md.

Example code coverage data on covedov.io:

https://codecov.io/gh/notmandatory/bdk/branch/codecov

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,6 +10,7 @@
   <p>
     <!-- <a href="https://crates.io/crates/magical"><img alt="Crate Info" src="https://img.shields.io/crates/v/magical.svg"/></a> -->
     <a href="https://github.com/bitcoindevkit/bdk/actions?query=workflow%3A%22Continuous+integration%22"><img alt="Continuous integration Status" src="https://github.com/bitcoindevkit/bdk/workflows/Continuous%20integration/badge.svg"></a>
+    <a href="https://codecov.io/gh/bitcoindevkit/bdk"><img src="https://codecov.io/gh/bitcoindevkit/bdk/branch/master/graph/badge.svg?token=5EG2G7GAM3"/></a>
```
Comment (afilini) [inline README.md]: Is that token necessary? If I try to open the url without it, it still seems to work fine

File: .github/workflows/code_coverage.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          override: true
+
+      - name: tarpaulin
+        uses: actions-rs/tarpaulin@v0.1.2
+        with:
+          version: '0.15.0'
+          args: '--exclude-files testutils -- --test-threads 1'
```
Comment (afilini) [inline .github/workflows/code_coverage.yml]: Ideally we should try to add as many features as we can here. I would add the `all-keys` feature to run a few more tests for the `keys` module, and if we can setup the environment with the docker image we could also run the full end to end electrum tests with the `test-electrum` feature.

Unfortunately most of the `blockchain` code (esplora, compact filters) will remain untested for now, at least until we are done with #102

**Comment by afilini** [comment]
I was looking at the report generated on your branch, it seems a bit dumb here that it doesn't ignore the code for wasm32 but we don't really have that many parts like this, so it should mostly be fine.

https://codecov.io/gh/notmandatory/bdk/src/codecov/src/wallet/time.rs

File: README.md (hunk 1 above)
Comment (notmandatory) [reply README.md]: This is just the URL provided on the settings page, it's not the same token used for uploading new test runs. But if it works without it then I'll take it off an we can try it that way. Worst case I just need to put it back on.

File: .github/workflows/code_coverage.yml (hunk 2 above)
Comment (notmandatory) [reply .github/workflows/code_coverage.yml]: There should be a way to combine the feature specific test runs into one on Codecov. I'll get all the features tested on the CI build first and get it added here too. Will let ya know when it's ready.

**Comment by notmandatory** [comment]
I think when I get all the feature tests combined into one result it should fix the `wasm32` and other similar gaps in testing.

File: README.md (hunk 1 above)
Comment (afilini) [reply README.md]: No problem then, I was afraid this was something related to a specific account/org id. If it was like that in the settings there's probably a valid reason, better to keep it i guess.

**Comment by notmandatory** [comment]
I've added a tarpaulin config file to run all the tests and doctests as a single command which seems to combine the test coverage into a single report. The downside is it's only running in a single thread and takes an hour to finish. But it's setup to only run when a PR is merged, so it shouldn't matter too much how long it takes. Anyone who want's a faster preview can run tarpaulin locally (on Linux) with: 

```shell
cargo install cargo-tarpaulin
cargo tarpaulin -o Html
```

**Comment by afilini** [comment]
Just out of curiosity, why does it need to run on a single thread? The only tests that need that are the end to end tests with electrs and bitcoind, but all the other tests in theory can be run in parallel

**Comment by afilini** [comment]
If we want to make this run faster I guess we could also try just making one build but with all the features we can enabled. Something like:

```
all-keys cli-utils compiler esplora key-value-db electrum esplora compact_filters
```

I can't think of any test that is **disabled** if there are some other features enabled. In theory having all of those enabled in one go should yield the same result as making separate reports and then combining them together.

For the CI it probably makes sense to test small subsets of features to make sure we aren't breaking anything in that regard, but for the code coverage I guess the goal should just be to get a report done, not really testing anything.

**Comment by notmandatory** [comment]
Good suggestion! I configured the job to run with all features in one big test run and it's much faster, done in about 22 mins. Coverage results look the same. It should be ready to go now.

https://codecov.io/gh/notmandatory/bdk/branch/codecov

**Comment by afilini** [comment]
For a future PR, if we want to make the CI and this code coverage job run a bit faster, we could try to install `librocksdb-dev` first using `apt-get`. In theory, when cargo tries to compile `librocksdb-sys` which is needed for the compact filters sync, it should find the library on the system and just link to it instead of recompiling it from scratch, which takes a few minutes.

**Pull request merged**